Fetch the remote notifier config

mmcdaris opened the pull request:
Adds support for fetching the remote notifier config when possible. The remote config feature creates a base for more flexible PHPBrake configuration and features at runtime. The remote config is cached for 10 minutes in a temp file in the system's temp directory. If fetching or caching fails or is unavailable for any reason, the default settings are used. 

A few notes on `Airbrake\Notifier` initialization options and the remote config:
- This Adds a new `remoteConfig` option to the initialization of the `Airbrake\Notifer` that defaults to `true`. If the `Airbrake\Notifier` is initialized with the `remoteConfig` set to `false`, the config will not be fetched from the remote and the default endpoint will be used.
- If the `host` option is specified with `Airbrake\Notifier` initialization that will take precedence over the host fetched from the remote config.

mmcdaris committed:
Add support for pulling remote notifier config

The remote notifier config is fetched and cached for 10 minutes.

mmcdaris committed:
Add errorNotification option to avoid sending

This will avoid sending errors to the service if the remote config
response specifies that errors are disabled.

mmcdaris committed:
Add remoteConfig option to the Notifier

When set to false the defaults are used and no remote config fetch is
attempted.

kyrylo commented on src/RemoteConfig.php:
```diff
@@ -0,0 +1,143 @@
+    protected $remoteConfigURL;
+    protected $httpClient;
+    protected $tempCache;
+    protected $tempCacheFilename = 'airbrake_cached_remote_config.json';
```
Why do we need to cache remote config? Couldn't we keep the config in memory?

mmcdaris replied:
I really wanted to do that as well. I searched a lot for a way to store data between or outside of requests without much luck. Everything that looked promising as a solution required PHP be compiled with an extra flag or installed with PECL. I couldn't really find anything I could depend on existing so I decided to use a temporary file to cache the remote config.

kyrylo commented on src/Notifier.php:
```diff
@@ -558,15 +562,21 @@ protected function errorNotifications()
     {
-        if (isset($this->errorConfig)) {
-            return $this->errorConfig;
+        if ($this->opt['remoteConfig'] == false) {
```
Will this work?

```suggestion
        if (!$this->opt['remoteConfig']) {

mmcdaris replied:
Yea this is nice. remote config will be either `true` or `false` so I'll add this in.

kyrylo commented on src/Notifier.php:
```diff
@@ -558,15 +562,21 @@ protected function errorNotifications()
         return $this->errorConfig;
     }
 
+    protected function newRemoteConfig($projectId)
```
Why does this have to be a function? Now we assign to `$remoteConfig` twice. I am not sure I am following this.

mmcdaris replied:
my main motivation for having a `newRemoteConfig` method is that it made things easier to mock in tests. 

We can remove the `$remoteConfig` assignment here though, sorry that was leftover from some local debugging. The body of this method should just be:
```php
   return new RemoteConfig($projectId);
```

Since the `$remoteConfig` variable isn't prefixed with `$this->` it will only be available in the method scope and not on the instance.

kyrylo reviewed (commented):
We need to fetch the config every 10 minutes (in background). Currently, unless I am mistaken, we fetch it only once, on launch. Did I read the code correctly?

mmcdaris committed:
use simpler condition for remoteConfig option check

Co-authored-by: Kyrylo Silin <silin@kyrylo.org>

mmcdaris committed:
Remove unused $remoteConfig variable from method

mmcdaris committed:
Use buildNoticesURL instead of NoticesURL

We don't need to set up or store the noticesURL during construction
since we get this information at the time the error occurs with the
buildNoticesURL function instead. This ensures we always get the
freshest error host to use from the RemoteConfig or when that isn't
available just fallback to the default error host of `api.airbrake.io`.

mmcdaris commented:
> We need to fetch the config every 10 minutes (in background). Currently, unless I am mistaken, we fetch it only once, on launch. Did I read the code correctly?

mmcdaris commented:
> We need to fetch the config every 10 minutes (in background). Currently, unless I am mistaken, we fetch it only once, on launch. Did I read the code correctly?

Hi @kyrylo thank you for the review and questions! 

I had another look at this an it turns out it was useless to set the `noticesURL` var in the constructor because we determine the URL when we report the error to Airbrake. I've tried to make this more clear in my [last commit](https://github.com/airbrake/phpbrake/pull/114/commits/a1a1ec17ec0df737631ee374c44fdf5c5000e94c) by removing the `noticesURL` var, and making it clear that we build the notice url whenever an error is reported. 

As for the cache lifecycle here's how it works: When an error is being reported to Airbrake, we ask the remoteConfig for the endpoint to send it to. If the remote config isn't cached or is expired, we fetch and cache it for the next 10 minutes. If the cache is still fresh we read from it. 

You mention this only getting fetched once on launch but as I understand it through my testing PHP doesn't hold onto instances of objects across requests and it would take extra effort to set up a global instance of `Airbrake\Notifier` that all errors are reported with. Even if that were the case the URL to use would be determined during the reporting of an error instead of during construction.

kyrylo commented on tests/NotifierTest.php:
```diff
@@ -2,6 +2,9 @@
 namespace Airbrake\Tests;
 
+// Update version across all tests in a single location.
+const AIRBRAKE_NOTIFIER_VERSION = '0.7.5';
```
Can't we read it from `src/Notifier.php`? This way we have less places to update the version.

mmcdaris replied:
Good idea, I've gotten the tests to pass using the namespace constant set in `src/Notifier.php`. 👍 for only setting the version in a single location.

kyrylo commented on src/RemoteConfig.php:
```diff
@@ -0,0 +1,143 @@
+    private function getConfigFromCacheOrFetch()
+    {
+        try {
+            if ($this->tempCache->canWrite() == false) {
```
```suggestion
            if (!$this->tempCache->canWrite()) {
```

kyrylo commented on src/RemoteConfig.php:
```diff
@@ -0,0 +1,143 @@
+    private function parseErrorConfig($config)
+    {
+        $config_not_found = array_key_exists("settings", (array) $config) == false;
+        if ($config_not_found) {
```
```suggestion
        $config_found = array_key_exists("settings", (array) $config);
        if (!$config_found) {
```

kyrylo commented on src/RemoteConfig.php:
```diff
@@ -0,0 +1,143 @@
+            return self::DEFAULT_CONFIG;
+        }
+
+        foreach ($config['settings'] as $cfg) {
```
In other notifiers I call `$cfg` as `$s` (short for setting). Not a big deal, but it would be more consistent with the rest of the notifier bros to call it `s` 😉

mmcdaris replied:
I like it and am here for the consistency. Change from `cfg->s` coming right up!

kyrylo commented on src/RemoteConfig.php:
```diff
@@ -0,0 +1,143 @@
+
+        foreach ($config['settings'] as $cfg) {
+            if (isset($cfg["name"]) && $cfg["name"] == "errors") {
+                $errorConfig = $cfg;
```
Once the setting is found, we can `break` immediately. No need continue iterating.

mmcdaris replied:
Good point, I've added the break.

kyrylo commented on src/RemoteConfig.php:
```diff
@@ -0,0 +1,143 @@
+            [
+                'connect_timeout' => 2,
+                'read_timeout' => 2,
+                'timeout' => 2,
```
Is `2` deliberate? Just checking, maybe you set it to 2 and meant to do something here later.

mmcdaris replied:
`2` was just a placeholder. The other client in Notifier.php uses `10` so I'll update to match that.

mmcdaris committed:
Use !condtion instead of condition == false

Co-authored-by: Kyrylo Silin <silin@kyrylo.org>

mmcdaris committed:
Assign and use config_found and if (!config_found)

Co-authored-by: Kyrylo Silin <silin@kyrylo.org>

mmcdaris committed:
Break the loop when the error settings are found

mmcdaris committed:
Use s for config setting loop variable

mmcdaris committed:
Match timeouts with the Notifier requests

mmcdaris committed:
Use Airbrake namespace version constant in tests

This makes it easy to update the version everywhere in only one place.

mmcdaris committed:
Memoize error config to avoid 2 cache calls

When we build an error we call errorHost() and errorNotifications(),
without memoization this would result in two calls to the cache instead
of just one.

mmcdaris committed:
Add remoteConfig option section to README
